add templating for pre defined app values

soer3n opened the pull request:
**What this PR does / why we need it**:
This pr adds the possiblity to use pre defined values in applicationdefinition and applicationinstallation resources for configuring the related helm release values.

**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #12530

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/kind chore
-->
/kind feature

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note. Release notes are being used to generate the changelog:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
It's now possible to use pre defined values in enterprise edition for applicationdefinition and applicationinstallation resources to specify cluster related values in helm release value configuration.
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
TBD
```

soer3n committed:
add templating for pre defined app values

* add fetching data for templating
* add parsing the data into the given helm values
* add unit tests for both new funcs
* add wrappers to make it ee only feature

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
fix helm integration tests and import order

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
add comment to function and fix log message

Signed-off-by: soer3n <srenhenning@googlemail.com>

xrstf commented on pkg/applications/providers/template/helm.go:
```diff
@@ -96,7 +96,13 @@ func (h HelmTemplate) InstallOrUpgrade(chartLoc string, appDefinition *appskuber
 	}
 
-	helmRelease, err := helmClient.InstallOrUpgrade(chartLoc, getReleaseName(applicationInstallation), values, *deployOpts, auth)
+	renderedValues, err := h.parsePreDefinedValues(values)
```
This function name is super confusing: Why do we have to parse the values we just got from `GetParsedValues()`? Maybe "templateValues" would be a more apt name?

soer3n replied:
Yes, you're right. I changed the function name to the suggested one `templatePreDefinedValues`

xrstf commented on pkg/applications/providers/template/helm.go:
```diff
@@ -96,7 +96,13 @@ func (h HelmTemplate) InstallOrUpgrade(chartLoc string, appDefinition *appskuber
+	renderedValues, err := h.parsePreDefinedValues(values)
+
+	if err != nil {
+		return util.NoStatusUpdate, fmt.Errorf("failed to render pre defined values: %w", err)
```
`pre-defined`

xrstf commented on pkg/applications/providers/template/wrappers_ce.go:
```diff
@@ -0,0 +1,23 @@
+//go:build !ee
+
+/*
+Copyright 2024 The Kubermatic Kubernetes Platform contributors.
```
2025 :grin:

xrstf commented on pkg/applications/providers/template/wrappers_ce.go:
```diff
@@ -0,0 +1,23 @@
+
+package template
+
+func (h *HelmTemplate) parsePreDefinedValues(applicationValues map[string]interface{}) (map[string]interface{}, error) {
```
optional nickpickery: `any` is shorter than `interface{}`

xrstf commented on pkg/applications/providers/template/wrappers_ee.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+func (h *HelmTemplate) parsePreDefinedValues(applicationValues map[string]interface{}) (map[string]interface{}, error) {
+	templateData, err := eeutil.GetTemplateData(h.Ctx, h.SeedClient, h.SecretNamespace)
```
I am super uncomfortable with this hidden "we deduce the Cluster from the HelmTemplate's secretNamespace" logic that is snuck in here. IMHO this should be deduced further up in the call stack by some controller and then the relevant Cluster object should be handed into the `InstallOrUpgrade()`.

Deducing clusters from namespaces has always worked and will continue to work, but it's just an ugly side effect that relies on knowing the internal behaviour of KKP. It's better to not write more code that relies on it.

soer3n replied:
The cluster name is now passed from the manager via applicationInstaller struct which is later used for the `InstallOrUpgrade()` function in the controller.

xrstf commented on pkg/applications/test/kubernetes.go:
```diff
@@ -65,6 +66,60 @@ func CreateNamespaceWithCleanup(t *testing.T, ctx context.Context, client ctrlru
+		},
+	}
+	if err := client.Create(ctx, cluster); err != nil {
+		t.Fatalf("failed to create test cluster: %s", err)
```
Please settle on `%v` for the placeholder, instead of switching between %v and %s.

xrstf commented on pkg/applications/test/kubernetes.go:
```diff
@@ -65,6 +66,60 @@ func CreateNamespaceWithCleanup(t *testing.T, ctx context.Context, client ctrlru
+		Versions: kubermaticv1.ClusterVersionsStatus{
+			ControlPlane: semver.Semver("v1.30.5"),
+		},
+		NamespaceName: "default",
```
Is it important that you deviate from KKP's logic (the logic your PR so much relies on)? If so, please leave a comment explaining why this is not "cluster-default". Otherwise, please change it to something realistic :grin:

soer3n replied:
This was actually a consequence of relying on the secret namespace. I changed this to `cluster-default` after changing the logic regarding passing the cluster name to the `InstallOrUpgrade` function.

xrstf commented on pkg/applications/test/kubernetes.go:
```diff
@@ -65,6 +66,60 @@ func CreateNamespaceWithCleanup(t *testing.T, ctx context.Context, client ctrlru
+	}
+
+	if err := client.Status().Update(ctx, cluster); err != nil {
+		t.Fatalf("failed to update testcluster status: %v", err)
```
`test cluster`

xrstf commented on pkg/applications/test/kubernetes.go:
```diff
@@ -65,6 +66,60 @@ func CreateNamespaceWithCleanup(t *testing.T, ctx context.Context, client ctrlru
+
+	t.Cleanup(func() {
+		if err := client.Delete(ctx, cluster); err != nil {
+			t.Fatalf("failed to cleanup test namespace: %s", err)
```
`cluster` instead of `namespace`

xrstf commented on pkg/applications/test/kubernetes.go:
```diff
@@ -65,6 +66,60 @@ func CreateNamespaceWithCleanup(t *testing.T, ctx context.Context, client ctrlru
+		}
+	})
+
+	if !utils.WaitFor(ctx, time.Second*1, time.Second*10, func() bool {
```
Since we're using a test client anyway, waiting for caches should not be necessary, no?

soer3n replied:
Yes, makes sense. I removed that.

xrstf commented on pkg/ee/applications/util.go:
```diff
@@ -0,0 +1,115 @@
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var ErrNoClusterForTemplating = fmt.Errorf("no cluster found for template data")
```
This can be `errors.New`

xrstf commented on pkg/applications/test/kubernetes.go:
```diff
@@ -65,6 +66,60 @@ func CreateNamespaceWithCleanup(t *testing.T, ctx context.Context, client ctrlru
+		c := &kubermaticv1.Cluster{}
+		return client.Get(ctx, ctrlruntimeclient.ObjectKeyFromObject(cluster), c) == nil
+	}) {
+		t.Fatalf("timeout waiting for namespace creation")
```
No need for `Fatalf`

soer3n replied:
I removed the whole timeout condition because it is not needed as mentioned by you in the above comment.

xrstf commented on pkg/ee/applications/util.go:
```diff
@@ -0,0 +1,115 @@
+
+var ErrNoClusterForTemplating = fmt.Errorf("no cluster found for template data")
+
+// TemplateData is the root context injected into each application values.
+type TemplateData struct {
+	Cluster ClusterData
+}
+
+// ClusterData contains data related to the user cluster
+// the application is rendered for.
+type ClusterData struct {
+	Name string
+	// HumanReadableName is the user-specified cluster name.
+	HumanReadableName string
+	// OwnerEmail is the owner's e-mail address.
+	OwnerEmail string
+	// ClusterAddress stores access and address information of a cluster.
+	Address kubermaticv1.ClusterAddress
+	// Version is the exact current cluster version.
+	Version string
+	// MajorMinorVersion is a shortcut for common testing on "Major.Minor" on the
+	// current cluster version.
+	MajorMinorVersion string
+}
```
This seems inspired by our addons and I think we need to document it in a similar fashion, otherwise it's impossible for users to make use of the variables. Checkout `codegen/godoc/` and `hack/ci/update-docs.sh` and extend as necessary.

soer3n replied:
I've added the generation of application template values and the current state. Thanks.

xrstf commented on pkg/ee/applications/util_test.go:
```diff
@@ -0,0 +1,197 @@
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var clusterNamespace = "test"
```
This also deviates from our namespace pattern.

soer3n replied:
Changed it to `test-cluster` to not deviate from patterns.

xrstf commented on pkg/ee/applications/util_test.go:
```diff
@@ -0,0 +1,197 @@
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := GetTemplateData(context.TODO(), tt.seedClient, tt.namespace)
```
Please use a Background context instead, or will there ever be a situation in the future where this context here will be replaced by something else? If not, no reason to leave a TODO for us that we will never work on. :grin:

xrstf commented on pkg/ee/applications/util_test.go:
```diff
@@ -0,0 +1,197 @@
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Fatalf("getTemplateData() got = %v, want %v", got, tt.want)
+			}
```
Checkout `pkg/test/diff` to make the error messages look nicer. A common pattern we do is

```
if changes := diff.Compare(......); changes != "" {
   Fatalf("Not the same\n%s", changes)
}
```

soer3n replied:
Thanks for the tip. I changed it for both tests. Returning errors and structs are now compared by `diff.ObjectDiff`.

xrstf commented on pkg/ee/applications/util_test.go:
```diff
@@ -0,0 +1,197 @@
+				},
+			},
+			want:    nil,
+			wantErr: fmt.Errorf("failed to render template: "),
```
Are errors really supposed to be like this? Just .... empty? Why is that?

soer3n replied:
I changed that to the complete error string while changing the function to compare returned values and expected in tests..

xrstf commented on pkg/ee/applications/util_test.go:
```diff
@@ -0,0 +1,197 @@
+			got, err := RenderValueTemplate(tt.values, &tt.templateData)
+			if (err != nil) && !strings.Contains(err.Error(), tt.wantErr.Error()) {
+				t.Fatalf("getTemplateData() error = %v, wantErr %v", err, tt.wantErr)
+			}
```
(sidenote; personally I am not a big fan of asserting error strings. It's brittle and can cause confusion when errors are reworded. If distinguishing between errors is crucial, errors should be dedicated types instead; in 99.9% of cases I personally in tests treat errors as quasi booleans that either happen or not)

soer3n replied:
I changed the comparison of return values in the added tests completely to the diff package. Thanks for the tip. I was not quite sure how to do this before.

Commit:
changes after review

* renaming function from parsePreDefinedValues to templatePreDefinedValues to be clearer
** change also map[string]interface{} to map[string]any for input parameters and return values
* fix some typos
* change year in new file headers to 2025
* pass clustername from contoller down to reconciler to be clearer when fetching the cluster resource for templating
** modify integration tests accordingly
* rename some values for testing to not deviate from kkp logic
* changes checking return errors in unit tests
** for TestGetTemplateData diff.StringDiff() will be used
** for TestRenderValueTemplate the type will be checked with `errors.As(...)`

Signed-off-by: soer3n <srenhenning@googlemail.com>

soer3n committed:
add godoc func for pre-defined app values + use internal funcs for test comparison

* add func to generate txt file with pre-defined values for templating to be used in docs
* use diff.CompareObject and diff.CompareString for error and struct comparison

Signed-off-by: soer3n <srenhenning@googlemail.com>

xrstf commented on pkg/ee/applications/util_test.go:
```diff
@@ -0,0 +1,201 @@
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := GetTemplateData(context.Background(), tt.seedClient, tt.namespace)
+			if diff.ObjectDiff(err, tt.wantErr) != "" {
+				t.Fatalf("GetTemplateData() diff: %s", diff.ObjectDiff(err, tt.wantErr))
```
ObjectDiff is not appropriate to compare errors, if comparing error strings is even desirable. ObjectDiff is useful if you have structured data that can be expressed as YAML.

soer3n replied:
I changed checking the expected errors to use either the type of the wrapped error or using a external function for checking a kube apiserver error.

xrstf commented on pkg/ee/applications/util_test.go:
```diff
@@ -0,0 +1,201 @@
+			}
+			if diff.ObjectDiff(got, tt.want) != "" {
+				t.Fatalf("GetTemplateData() got = %v, want %v", got, tt.want)
+			}
```
You're already using a _diff_ function, why not make use of the generated diff?

```
if changes := diff.ObjectDiff(got, tt.want); changes != "" {
   t.Fatalf("Received unexpected result:\n%s", changes)
}
```

xrstf commented on pkg/ee/applications/util_test.go:
```diff
@@ -0,0 +1,201 @@
+	clusterNamespace           = "test-cluster"
+	ErrNoClusterForTemplating  = errors.New("failed to get cluster \"test-cluster\": clusters.kubermatic.k8c.io \"test-cluster\" not found")
+	ErrNoValidClusterVersion   = errors.New("failed to parse semver version for cluster \"test-cluster\"")
+	ErrUnknownTemplateVariable = errors.New("failed to render template: template: application-pre-defined-values:2:14: executing \"application-pre-defined-values\" at <.Foo.Name>: can't evaluate field Foo in type *applications.TemplateData")
```
These errors don't need to be exported, do they?

soer3n replied:
These vars are completely removed because checking expected errors is now different like discussed before.

xrstf commented on pkg/applications/test/kubernetes.go:
```diff
@@ -65,6 +66,53 @@ func CreateNamespaceWithCleanup(t *testing.T, ctx context.Context, client ctrlru
+
+	t.Cleanup(func() {
+		if err := client.Delete(ctx, cluster); err != nil {
+			t.Fatalf("failed to cleanup test cluster: %s", err)
```
%v for consistency

soer3n committed:
changes after review

* change error messages for testing to unexported vars
* modify error messages in unit tests
* changes error formatting from %s to %v

Signed-off-by: soer3n <srenhenning@googlemail.com>

xrstf commented on pkg/ee/applications/util_test.go:
```diff
@@ -0,0 +1,201 @@
+
+			if diff.ObjectDiff(got, tt.want) != "" {
+				t.Fatalf("RenderValueTemplate() got = %v, want %v", got, tt.want)
+			}
```
Same diff confusion 😁
